core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -0,0 +1,340 @@
+import static org.mule.runtime.api.functional.Either.right;
+import static org.mule.runtime.api.metadata.DataType.fromObject;
+import static org.mule.runtime.core.api.event.CoreEvent.builder;
+import static org.mule.runtime.core.internal.routing.ExpressionSplittingStrategy.DEFAULT_SPLIT_EXPRESSION;
+import static org.mule.runtime.core.privileged.processor.MessageProcessors.applyWithChildContext;
+import static org.slf4j.LoggerFactory.getLogger;
+import static reactor.core.Exceptions.propagate;
+import static reactor.core.publisher.Mono.subscriberContext;
+import static reactor.util.context.Context.empty;
+
+class ForeachRouter {
+
+  private static final Logger LOGGER = getLogger(ForeachRouter.class);
+
+  private static final String MULE_FOREACH_CONTEXT_KEY = "mule.foreach.router.foreachContext";
+  static final String MAP_NOT_SUPPORTED_MESSAGE =
+      "Foreach does not support 'java.util.Map' with no collection expression. To iterate over Map entries use '#[dw::core::Objects::entrySet(payload)]'";
+  private final Map<String, ForeachContext> foreachContextResolver = new HashMap<>();
+  private final Foreach owner;
+
+  private Flux<CoreEvent> upstreamFlux;
+  private Flux<CoreEvent> innerFlux;
+  private Flux<CoreEvent> downstreamFlux;
+  private final FluxSinkRecorder<CoreEvent> innerRecorder = new FluxSinkRecorder<>();
+  private final FluxSinkRecorder<Either<Throwable, CoreEvent>> downstreamRecorder = new FluxSinkRecorder<>();
+  private final AtomicReference<Context> downstreamCtxReference = new AtomicReference(empty());
+
+  private final AtomicInteger inflightEvents = new AtomicInteger(0);
+  private final AtomicBoolean completeDeferred = new AtomicBoolean(false);
+
+  ForeachRouter(Foreach owner, Publisher<CoreEvent> publisher, String expression, int batchSize,
+                MessageProcessorChain nestedChain) {
+    this.owner = owner;
+
+    upstreamFlux = Flux.from(publisher)
+        .doOnNext(event -> {
+
+          if (expression.equals(DEFAULT_SPLIT_EXPRESSION)
+              && Map.class.isAssignableFrom(event.getMessage().getPayload().getDataType().getType())) {
+            downstreamRecorder.next(left(new IllegalArgumentException(MAP_NOT_SUPPORTED_MESSAGE)));
+          }
+
+          final CoreEvent responseEvent =
+              builder(event).addVariable(owner.getRootMessageVariableName(), event.getMessage()).build();
+
+          // Create ForEachContext
+          ForeachContext foreachContext = this.createForeachContext(event);
+
+          // Save it to handle nested routers
+          foreachContextResolver.put(event.getContext().getId(), foreachContext);
+
+          // Save it inside the internalParameters of the event
+          ((InternalEvent) responseEvent).setForEachContext(foreachContext);
+
+          inflightEvents.getAndIncrement();
+          try {
+            // Set the Iterator<TypedValue> calculated
+            foreachContext.setIterator(this.splitRequest(responseEvent, expression));
```

Comment:
try ti do the split early, since it may fail. Also, the split logic should be provided by the component.

Reply:
the call to `foreachContext.setIterator` should go inside `prepareEvent` too.